File: go/BinaryTreePath/BinaryTreePath.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func binaryTreePaths(root *TreeNode) []string {
	result := make([]string, 0)
	recurse(root, "", &result)
	return result
}

// Time -> NlogN Space -> NlogN
func recurse(root *TreeNode, path string, result *[]string) {
	if root == nil {
		return
	}

	if root.Left == nil && root.Right == nil {
		*result = append(*result, path+fmt.Sprintf("%v", root.Val))
		return
	}

	path += fmt.Sprintf("%v", root.Val) + "->"

	recurse(root.Left, path, result)
	recurse(root.Right, path, result)
}

File: python/binaryTreePath.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def binaryTreePaths(root: TreeNode | None) -> list[str]:
    result = []
    recurse(root, "", result)
    return result


# Time -> NlogN Space -> NlogN
def recurse(root: TreeNode | None, path: str, result: list[str]):
    if root is None:
        return
    if root.left is None and root.right is None:
        result.append(path + str(root.val))
        return

    path += str(root.val) + "->"
    recurse(root.left, path, result)
    recurse(root.right, path, result)
